sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,729 @@
+allocations from the memory pool in this extension be used for the bindless
+images, however, this will need to be investigated further and so the details of
+this are not included in this version of the extension.
+
+== Examples
+
+There are various ways in which this extension can be used but a typical usage
+of the memory pool and the asynchronous malloc and free commands is described
+below. In this example an explicit memory pool is created and this is used to
+share memory allocated from the OS between multiple asynchornous malloc
+commands.
+
+[source,c++]
+----
+int main(int argc, char *argv[])
+{
+  queue q(property::queue::in_order);
+
+  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
+    q.get_device(), usm::alloc::device);
+  
+  {
+    // memory pool allocates memory from the OS
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for first kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  {
+    // memory pool re-uses previously allocated memory
+    void *temp = async_malloc_from_pool(q, 1024, memPool);
+
+    // memory allocation is used for second kernel
+    parallel_for(q, range{1024}, [=](id<1> idx) {
+      do_something_else(idx, temp);
+    });
+
+    // memory is available to be used by another allocation
+    async_free(q, temp);
+  }
+
+  // memory pool releases memory back to the OS
+  q.wait();
+}
+----
+
+The above example does not use SYCL events, so below is the same example using
+an out-of-order SYCL queue and SYCL events to manage dependencies.
+
+[source,c++]
+----
+int main(int argc, char *argv[])
+{
+  queue q;
+
+  ext::oneapi::experimental::memory_pool memPool(q.get_context(),
+    q.get_device(), usm::alloc::device);
+  
+  {
+    void *temp = null;
+
+    // memory pool allocates memory from the OS
+    auto e1 = q.submit_with_event([&](handler &cgh) {
+      temp = async_malloc_from_pool(q, 1024, memPool);
+    });
+
+    // memory allocation is used for first kernel
+    auto e2 = q.submit_with_event([&](handler &cgh) {
+      cgh.depends_on(e1);
+      parallel_for(q, range{1024}, [=](id<1> idx) {
+        do_something(idx, temp);
+      });
+    });
+
+    // memory is available to be used by another allocation
+    auto e3 = q.submit_with_event([&](handler &cgh) {
+      cgh.depends_on(e2);
+      async_free(q, temp);
+    });
+  }
+
+  {
+    void *temp = null;
+
+    // memory pool re-uses previously allocated memory
+    auto e1 = q.submit_with_event([&](handler &cgh) {
+      temp = async_malloc_from_pool(q, 1024, memPool);
+    });
+
+    // memory allocation is used for second kernel
+    auto e2 = q.submit_with_event([&](handler &cgh) {
+      cgh.depends_on(e1);
+      parallel_for(q, range{1024}, [=](id<1> idx) {
+        do_something_else(idx, temp);
```

Comment:
Not sure if that works as expected with a just plain pointer. `temp` is captured by value by the kernel function, but there is no guarantee that the final value is available. Because the preceding allocation command is supposed to be executed asynchronously.

Reply:
I also have the concern, the 'first 3 submit_with_event' and the 'last 3 submit_with_event' are independent, if we still have hardware resource available when do_something is executed, it is possible that do_something_else is scheduled to execution concurrently.

Reply:
looks that 74d6487 does not change meaningful code.

> there should be dependency between the first async_free and the second async_malloc_from_pool
why there's dependency between them?

Reply:
Good catch I missed the `depends_on` in the fourth command group, fixed in https://github.com/intel/llvm/pull/14800/commits/9c2cacf9fb07b96f11014d48359891e19545ce93.